lib/upgradable-multisig/multisig/utils.ak

```diff
@@ -0,0 +1,62 @@
+use aiken/list
+use aiken/transaction.{Input, Output}
+use aiken/transaction/credential.{Credential}
+use aiken/transaction/value.{PolicyId, Value}
+use upgradable_multisig/types.{AssetClass, MultisigDatum, PubKeyHash}
+
+pub fn get_input_by_token(inputs: List<Input>, policy_id: PolicyId) {
+  list.find(
+    inputs,
+    fn(input) { list.has(value.policies(input.output.value), policy_id) },
+  )
+}
+
+pub fn validate_multisig_datum(datum: MultisigDatum) -> Bool {
+  (list.length(datum.signers) > 1)? && (datum.threshold > 0)? && (datum.spending_limit > 2_000_000)?
+}
+
+// Helper function to get the amount of a specific asset from a Value
+pub fn get_asset_amount(value: Value, asset: AssetClass) -> Int {
+  value.quantity_of(value, asset.policy_id, asset.asset_name)
+}
+
+// Helper function to validate where the NFT goes
+pub fn validate_token_output(
+  output: Output,
+  policy: PolicyId,
+  token_name: ByteArray,
+  credential: Credential,
+  is_credential: Bool,
+) -> Bool {
+  let token_quantity = value.quantity_of(output.value, policy, token_name)
+  let payment_credential = output.address.payment_credential
+
+  token_quantity == 1 && if is_credential {
+    payment_credential == credential
+  } else {
+    payment_credential != credential
+  }
+}
+
+pub fn signed_within_threshold(
+  datum: MultisigDatum,
+  extra_signatories: List<PubKeyHash>,
+) -> Bool {
+  let signer_count_length =
+    list.foldr(
+      extra_signatories,
+      0,
+      fn(current_signer, accumulator) -> Int {
+        if list.has(datum.signers, current_signer) {
+          accumulator + 1
+        } else {
+          accumulator
+        }
+      },
+    )
```

Comment:
The foldr is less effective then foldl. Also, I suggest creating a stoppable recursive function for this, for example:
```
fn has_enough_signers(all_signers: List<PubKeyHash>, threshold: Int, actual_signers: List<PubKeyHash>) -> Bool {
  when all_singers is {
    [] -> False
    [head, ..tail] ->
      if list.has(actual_signers, head) {
        if threshold > 1 {
          has_enough_signers(tail, threshold - 1, actual_signers)
        } else {
          True
        }
      } else {
        has_enough_signers(tail, threshold, actual_signers)
      }
  }
}

pub fn signed_within_threshold(
  datum: MultisigDatum,
  extra_signatories: List<PubKeyHash>,
) -> Bool {
  has_enough_signers(datum.signers, datum.threshold, extra_signatories)
}
```